drivers/flash/flash_ite_it8xxx2.c

```diff
@@ -0,0 +1,582 @@
+					  off_t offset, size_t len)
+{
+	struct flash_it8xxx2_dev_data *data = DEV_DATA(dev);
+	int v_size = len, v_addr = offset, ret = -EINVAL;
+	unsigned int key;
+
+	/*
+	 * Check that the offset and length are multiples of the write
+	 * erase block size.
+	 */
+	if ((offset % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+	if ((len % FLASH_ERASE_BLK_SZ) != 0) {
+		return -EINVAL;
+	}
+	if (data->flash_static_cache_enabled == 0) {
+		return -EACCES;
+	}
+	if (data->all_protected) {
+		return -EACCES;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+	/*
+	 * CPU can't fetch instruction from flash while use
+	 * EC-indirect follow mode to access flash, interrupts need to be
+	 * disabled.
+	 */
+	key = irq_lock();
+
+	/* Always use sector erase command */
+	for (; len > 0; len -= FLASH_ERASE_BLK_SZ) {
+		ramcode_flash_erase(offset, FLASH_CMD_SECTOR_ERASE);
+		offset += FLASH_ERASE_BLK_SZ;
+	}
+	ramcode_reset_i_cache();
+	/* get the ILM address of a flash offset. */
+	v_addr |= CHIP_MAPPED_STORAGE_BASE;
+	ret = ramcode_flash_verify(v_addr, v_size, NULL);
+
+	irq_unlock(key);
+
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+/* Enable or disable the write protection */
+static int flash_it8xxx2_write_protection(const struct device *dev,
+					  bool enable)
+{
+	struct flash_it8xxx2_dev_data *data = DEV_DATA(dev);
+
+	if (enable) {
+		/* Protect the entire flash */
+		flash_protect_banks(0, CHIP_FLASH_SIZE_BYTES /
+			CHIP_FLASH_BANK_SIZE, FLASH_WP_EC);
+		data->all_protected = 1;
+	} else {
+		data->all_protected = 0;
+	}
+
+	/*
+	 * bit[0], eflash protect lock register which can only be write 1 and
+	 * only be cleared by power-on reset.
+	 */
+	IT83XX_GCTRL_EPLR |= 0x01;
+
+	return 0;
+}
+
+static const struct flash_parameters *
+flash_it8xxx2_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_it8xxx2_parameters;
+}
+
+static void flash_code_static_cache(const struct device *dev)
+{
+	struct flash_it8xxx2_regs *const flash_regs = FLASH_IT8XXX2_REG_BASE;
+	struct flash_it8xxx2_dev_data *data = DEV_DATA(dev);
+	unsigned int key;
+
+	/* Make sure no interrupt while enable static cache */
+	key = irq_lock();
+
+	/* invalid static cache first */
+	IT83XX_GCTRL_RVILMCR0 &= ~ILMCR_ILM0_ENABLE;
+
+	flash_regs->SMFI_SCAR0H = BIT(3);
+
+	memcpy((void *)CHIP_RAMCODE_BASE, (const void *)FLASH_RAMCODE_START,
+		IT8XXX2_ILM_BLOCK_SIZE);
+
+	/* RISCV ILM 0 Enable */
+	IT83XX_GCTRL_RVILMCR0 |= ILMCR_ILM0_ENABLE;
+
+	/* Enable ILM */
+	flash_regs->SMFI_SCAR0L = FLASH_RAMCODE_START & 0xFF;
+	flash_regs->SMFI_SCAR0M = (FLASH_RAMCODE_START >> 8) & 0xFF;
+	flash_regs->SMFI_SCAR0H = (FLASH_RAMCODE_START >> 16) & 0x7;
+
+	if (FLASH_RAMCODE_START & BIT(19)) {
+		flash_regs->SMFI_SCAR0H |= BIT(7);
+	} else {
+		flash_regs->SMFI_SCAR0H &= ~BIT(7);
+	}
+	/*
+	 * Validate Direct-map SRAM function by programming
+	 * register SCARx bit20=0
+	 */
+	flash_regs->SMFI_SCAR0H &= ~BIT(4);
+
```

Comment:
Lines 516-538 include BIT(x) magix values. Can you please create readable defines to make the code easier to read?

Reply:
Done